File: Hook/HookSkinDemo/app/src/main/java/com/dream/hookskindemo/MainActivity.java
package com.dream.hookskindemo;

import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;

import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.view.View;

import com.dream.hookskindemo.base.BaseActivity;

public class MainActivity extends BaseActivity {

    private static final int REQUEST_EXTERNAL_STORAGE = 1;
    private static String[] PERMISSIONS_STORAGE = {
            "android.permission.READ_EXTERNAL_STORAGE",
            "android.permission.WRITE_EXTERNAL_STORAGE"};//内存读写的权限，现在要动态申请了？

    /**
     * 申请权限，为了要把外部文件写入到 手机内存中
     *
     * @param activity
     */
    public static void verifyStoragePermissions(AppCompatActivity activity) {

        try {
            //检测是否有写的权限
            int permission = ActivityCompat.checkSelfPermission(activity,PERMISSIONS_STORAGE[1]);
            if (permission != PackageManager.PERMISSION_GRANTED) {
                // 没有写的权限，去申请写的权限，会弹出对话框
                ActivityCompat.requestPermissions(activity, PERMISSIONS_STORAGE, REQUEST_EXTERNAL_STORAGE);
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);//执行父类的onCreate，父类里面干了什么？看吧
        setContentView(R.layout.activity_main);//常规操作，setContentView


        verifyStoragePermissions(this);//申请权限。。
        findViewById(R.id.button).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                changeSkin(MainActivity.this,getPath());
            }
        });
        findViewById(R.id.button2).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent intent = new Intent(MainActivity.this, SecondActivity.class);
                startActivity(intent);
            }
        });
    }


}

File: Algorithms/UIEffectDemo/app/src/main/java/com/dream/uieffectdemo/MainActivity.kt
package com.dream.uieffectdemo

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import androidx.databinding.DataBindingUtil
import androidx.lifecycle.ViewModelProvider
import com.blankj.utilcode.util.GsonUtils
import com.blankj.utilcode.util.JsonUtils
import com.dream.uieffectdemo.baservadapterhelper.BaseRvAdapterHelperActivity
import com.dream.uieffectdemo.databinding.ActivityMainBinding
import com.dream.uieffectdemo.imedittext.IMEditTextActivity
import com.dream.uieffectdemo.popupwindow.PopupWindowActivity
import com.dream.uieffectdemo.text.TextActivity
import com.dream.uieffectdemo.utils.JsonFileReader


class MainActivity : AppCompatActivity() {

    private var mBinding: ActivityMainBinding? = null
    private var mViewModel: MainViewModel? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        //获取 dataBinding
        mBinding = DataBindingUtil.setContentView(this, R.layout.activity_main)
        //获取 ViewModel
        mViewModel = ViewModelProvider(this).get(MainViewModel::class.java)
        mBinding?.lifecycleOwner = this
        mBinding?.viewModel = mViewModel
        val json = JsonFileReader.getJson(this, "myjson.json")
        Log.d("erdai", "onCreate: $json")
        val text = JsonUtils.getString(json,"text")
        val formatStringByTemplate = getFormatStringByTemplate(text, json)
        Log.d("erdai", "onCreate: $formatStringByTemplate")


        val str = "*****\${queueNum}***".replace("replace","12")
        Log.d("erdai", "onCreate: $str")
        Log.d("erdai", "onCreate: ${"str".plus("哈哈哈").plus("*")}")
        initClickListener()
    }

    @Suppress("UNCHECKED_CAST")
    fun getFormatStringByTemplate(sourceString: String, template: String): String{
        try {
            val templateMap: Map<String,Any?> = GsonUtils.fromJson(template,Map::class.java) as Map<String, Any?>
            var formatString = sourceString
            for((key,value) in templateMap){
                val replace = "\${key}".replace("key", key)
                formatString = formatString.replace(replace,value?.toString()?:"")
            }
            return formatString
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return ""
    }

    private fun initClickListener() {
        //popWindow 点击
        mBinding?.popupWindowClick = View.OnClickListener{
            startActivityInner<PopupWindowActivity>(this)
        }

        //smartEditText 点击
        mBinding?.smartEditTextClick = View.OnClickListener {
            startActivityInner<IMEditTextActivity>(this)
        }

        mViewModel?.baseQuickAdapterClick = View.OnClickListener {
            startActivityInner<BaseRvAdapterHelperActivity>(this)
        }

        mBinding?.textClick = View.OnClickListener {
            startActivityInner<TextActivity>(this)
        }
    }



    companion object{
        inline fun <reified T> startActivityInner(context: Context){
            val intent = Intent(context,T::class.java)
            context.startActivity(intent)
        }
    }

}